### ISSUE TITLE
Option to create CloudFront distribution on media package create channel

### ISSUE THREAD

**Issue by samueleastdev** [aws/aws-sdk-php#1656]
When you create a media package channel via the php sdk there is no option to create a CloudFront distribution you need to go to the channel and manually click create CloudFront distribution.

https://docs.aws.amazon.com/aws-sdk-php/v3/api/api-mediapackage-2017-10-12.html#createchannel

![screen shot 2018-10-23 at 20 55 10](https://user-images.githubusercontent.com/1894778/47387024-1bbdcf80-d706-11e8-8840-7cf2ac3e4a3a.png)
![screen shot 2018-10-23 at 20 54 45](https://user-images.githubusercontent.com/1894778/47387034-224c4700-d706-11e8-9fab-13ad7657c52a.png)

It would be good if there was an option to set this up upon creation.

**Comment by diehlaws** [comment]
Hi @samueleastdev, thanks for reaching out to us. The feature you are describing is a customization of the AWS Console that is not available for any of the AWS SDKs; the [`CreateChannel`](https://docs.aws.amazon.com/mediapackage/latest/apireference/channels.html) API for MediaPackage does not include an option to automatically create a CloudFront distribution upon channel creation. 

In order to create a CloudFront distribution for your newly created MediaPackage channel you will first need to create an endpoint for your channel with a `CreateOriginEndpoint` call, then issue a CloudFront [`CreateDistribution`](https://docs.aws.amazon.com/aws-sdk-php/v3/api/api-cloudfront-2018-06-18.html#createdistribution) call using the endpoint you created for the channel as the distribution's origin.

If you would like to see this console feature implemented as part of the MediaPackage API, I suggest submitting a service feature request by creating a [new support case](https://console.aws.amazon.com/support/v1#/case/create) under the MediaPackage service, or posting in the [AWS Elemental MediaPackage Forums](https://forums.aws.amazon.com/forum.jspa?forumID=261).

**Comment by khizarbajwa007** [comment]
Created MediaLive input -> then MediaPackage Channel (Dash and Hsl) endpoint and lastly created Medialive channel

now I want to provide Dash and Hls to CloudFront but the documentation is not helping me out

**Comment by sirdeggen** [comment]
@khizarbajwa007 I'm having the _exact_ same issue. For now I've been trying to use the CloudFront aws-sdk to attempt to create a distribution after the endpoints are added but to no avail.

Keep getting:

`InvalidOrigin: The specified origin server does not exist or is not valid.`

```
import { MediaLive, MediaPackage, CloudFront } from 'aws-sdk'

const cf = new CloudFront({
    region: process.env.region,
    accessKeyId: process.env.accessKeyId,
    secretAccessKey: process.env.secretAccessKey
})

const newNumber = '2' // new channel id

cf.createDistribution({
            DistributionConfig: { /* required */
                CallerReference: new Date().toISOString(), /* required */
                Comment: 'Managed by MediaPackage_Channel_' + newNumber, /* required */
                DefaultCacheBehavior: { /* required */
                    ForwardedValues: { /* required */
                        Cookies: { /* required */
                            Forward: 'all', /* required */
                        },
                        QueryString: true, /* required */
                        Headers: {
                            Quantity: '0' /* required */
                        },
                        QueryStringCacheKeys: {
                            Quantity: '0' /* required */
                        }
                    },
                    MinTTL: '6', /* required */
                    TargetOriginId: "Endpoint_DASH_" + newNumber, /* required */
                    TrustedSigners: { /* required */
                        Enabled: false, /* required */
                        Quantity: '0' /* required */
                    },
                    ViewerProtocolPolicy: 'redirect-to-https', /* required */
                    AllowedMethods: {
                        Items: [ /* required */
                            'GET', 'HEAD', 'OPTIONS'
                        ],
                        Quantity: '3', /* required */
                    },
                    MaxTTL: '60'
                },
                Enabled: true, /* required */
                Origins: { /* required */
                    Items: [ /* required */
                        {
                            DomainName: originDomainName, /* required */
                            Id: "Endpoint_DASH_" + newNumber, /* required */
                        }
                    ],
                    Quantity: '1' /* required */
                },
                PriceClass: 'PriceClass_All'
            }
        }, (err, data) => console.log(err, data))
```

**Comment by samueleastdev** [comment]
Thanks for this helped me out, though I would post a working php version for anyone interested.

~~~
try {

 	
    	$callerReference = md5(uniqid(rand(), true));

		$comment         = 'Media Package Streaming';

		$cloudfrontClient = new CloudFrontClient([
		    'version'     => 'latest',
		    'region'      => 'us-west-2',
		    'credentials' => [],
		]);

		$createDistribution = $cloudfrontClient->createDistribution([
		    'DistributionConfig' => [ 
		        'Aliases' => [
		            'Quantity' => 0, 
		        ],
		        'CallerReference' => $callerReference, 
		        'Comment' => $comment, 
		        'CustomErrorResponses' => [
		            'Quantity' => 0, 
		        ],
		        'DefaultCacheBehavior' => [ 
		        	'TargetOriginId' => 'EMP-random-string', 
		        	'AllowedMethods' => [
	                    'Items' => ["GET", "HEAD"],
	                    'Quantity' => 2,
	                    'CachedMethods' => [
		                    'Items' => ["GET", "HEAD"],
		                    'Quantity' => 2
		                ],
	                ],
		            'Compress' => false,
		            'ForwardedValues' => [ 
		                'Cookies' => [ 
		                    'Forward' => 'none', 
		                    'WhitelistedNames' => [
		                        'Quantity' => 0, 
		                    ],
		                ],
		                'Headers' => [ 
		                    'Quantity' => 0, 
		                ],
                        'QueryStringCacheKeys' => [
                            'Items' => ['end', 'm', 'start'],
                            'Quantity' => 3, 
                        ],
		                'QueryString' => true, 
		            ],
		            'MinTTL' => 3600, 
		            'SmoothStreaming' => false,
		            'TrustedSigners' => [ 
		                'Enabled' => false, 
		                'Quantity' => 0, 
		            ],
		            'ViewerProtocolPolicy' => 'redirect-to-https'
		        ],
		        'Enabled' => true, 
		        'IsIPV6Enabled' => false,
		        'Logging' => [
		            'Bucket' => '', 
		            'Enabled' => false, 
		            'IncludeCookies' => false, 
		            'Prefix' => '', 
		        ],
		        'Origins' => [ 
		            'Items' => [
		                [
		                	'Id' => 'EMP-random-string', 
		                    'DomainName' => '00000000.mediapackage.us-east-1.amazonaws.com',     
		                    'CustomHeaders' => [
		                        'Quantity' => 0
		                    ],
		                    'ConnectionAttempts' => 3,
                    		'ConnectionTimeout' => 10,
		                    'CustomOriginConfig' => [
		                        'HTTPPort' => 80, 
		                        'HTTPSPort' => 443, 
		                        'OriginKeepaliveTimeout' => 5,
		                        'OriginProtocolPolicy' => 'match-viewer', 
		                        'OriginReadTimeout' => 30
		                    ]
		                ],
		                [
		                	'Id' => 'TEMP_ORIGIN_ID/channel_1', 
		                    'DomainName' => 'mediapackage.amazonaws.com', 
		                    'CustomHeaders' => [
		                        'Quantity' => 0
		                    ],
		                    'ConnectionAttempts' => 3,
                    		'ConnectionTimeout' => 10,
		                    'CustomOriginConfig' => [
		                        'HTTPPort' => 80, 
		                        'HTTPSPort' => 443, 
		                        'OriginKeepaliveTimeout' => 5,
		                        'OriginProtocolPolicy' => 'match-viewer', 
		                        'OriginReadTimeout' => 30
		                    ]
		                ]
		            ],
		            'Quantity' => 2, 
		        ],
		        'OriginGroups' => [
		            'Quantity' => 0
		        ],
		        'PriceClass' => 'PriceClass_All',
		    ],
		])->toArray();

    } catch (AwsException $e) {

        wp_send_json(array(
            'status' => false,
            'error' => $e->getMessage()
        ));

    }
~~~

**Comment by pgbce** [comment]
Hello, I've been able to get as far as creating the MediaPackageChannel, Endpoint and creating the cloudfront distribution with the configurations based on what was provided here. 
Unfortunately, MediaPackage Endpoint still has no Cloudfront URL filled out.

Any potential solutions to this? I've also scoured the documentation and there doesn't seem to be any update channel or update endpoint functionality to include CloudfrontURL.

<img width="624" alt="Screen Shot 2021-11-04 at 4 57 02 PM" src="https://user-images.githubusercontent.com/70154261/140418933-31021186-6bc7-420d-a17d-689775911efa.png">
